core/src/validation.jl

```diff
@@ -406,59 +406,82 @@ function sorted_table!(
     return table
 end
 
+struct NodeID
+    value::Int
+end
+
+Base.broadcastable(id::NodeID) = Ref(id)
+Base.show(io::IO, id::NodeID) = print(io, "#$(id.value)")
+
+function Base.isless(id_1::NodeID, id_2::NodeID)::Bool
+    return id_1.value < id_2.value
+end
+
+function Base.getindex(M::AbstractArray, id_row::NodeID, id_col::NodeID)
+    return M[id_row.value, id_col.value]
+end
+
+function Base.setindex!(
+    M::AbstractArray,
+    value::T,
+    id_row::NodeID,
+    id_col::NodeID,
+)::Nothing where {T}
+    M[id_row.value, id_col.value] = value
+    return nothing
+end
+
+struct EdgeID
+    value::Int
+end
+
 """
 Test for each node given its node type whether the nodes that
 # are downstream ('down-edge') of this node are of an allowed type
 """
-function valid_edges(
-    edge_ids::Dictionary{Tuple{Int, Int}, Int},
-    edge_connection_types::Dictionary{Int, Tuple{Symbol, Symbol}},
-)::Bool
-    rev_edge_ids = dictionary((v => k for (k, v) in pairs(edge_ids)))
-    errors = String[]
-    for (edge_id, (from_type, to_type)) in pairs(edge_connection_types)
-        if !(to_type in neighbortypes(from_type))
-            a, b = rev_edge_ids[edge_id]
-            push!(
-                errors,
-                "Cannot connect a $from_type to a $to_type (edge #$edge_id from node #$a to #$b).",
-            )
+function valid_edges(graph::MetaGraph)::Bool
+    errors = false
+    for e in edges(graph)
+        id_src = label_for(graph, e.src)
+        id_dst = label_for(graph, e.dst)
+        type_src = graph[id_src].type
+        type_dst = graph[id_dst].type
+
+        if !(type_dst in neighbortypes(type_src))
+            errors = true
+            edge_id = graph[id_src, id_dst].id.value
+            @error "Cannot connect a $type_src to a $type_dst (edge #$edge_id from node $id_src to $id_dst)."
```

Comment:
If you would define a `show` method for EdgeID as well, you wouldn't have to take the `.value` out here right?

Perhaps it's worth it to create an `abstract type ID`, and make both NodeID and EdgeID an ID, and then you can directly define these methods for both.